### ISSUE TITLE
false dependency between instance constraints

### ISSUE THREAD

**Issue by davdar** [agda/agda#1955]
This file doesn't typecheck because there is no instance for `Y unit` (which is commented out). The issue is the error, which reports:

"No instance of type Z unit was found in scope."

This appears to be an off-by-one error in reporting errors when instance resolution fails; the correct error output should identify `Y unit` as the missing instance.

Extra notes: if the `ZA` constraint on foo is removed, it gets the error message right claiming it failed to find `Y unit`. If the `XA` constraint is removed, it continues to give the incorrect error which mentions `Z unit`.

``` agda
module Bug where

data unit : Set where tt : unit

record X (A : Set) : Set where field x : A
record Y (A : Set) : Set where field y : A
record Z (A : Set) : Set where field z : A

instance
  X[unit] : X unit
  X[unit] = record { x = tt }
  -- Y[unit] : Y unit
  -- Y[unit] = record { y = tt }
  Z[unit] : Z unit
  Z[unit] = record { z = tt }

foo : ∀ (A : Set) {{XA : X A}} {{YA : Y A}} {{ZA : Z A}} → unit
foo A = tt

foo[unit] : unit
foo[unit] = foo unit
```

**Comment by andreasabel** [comment]
Confirmed. You can drop `X` from the picture.

``` agda
data unit : Set where tt : unit

record Y (A : Set) : Set where field y : A
record Z (A : Set) : Set where field z : A

instance
  -- Y[unit] : Y unit
  -- Y[unit] = record { y = tt }
  Z[unit] : Z unit
  Z[unit] = record { z = tt }

foo : ∀ (A : Set) {{YA : Y A}} {{ZA : Z A}} → unit
foo A = tt

foo[unit] : unit
foo[unit] = foo unit -- {{ZA = Z[unit]}}
```

If we give the instance for `ZA` explicitly, the error is correct.

**Comment by andreasabel** [comment]
This worked fine up to Agda-2.4.2.6.  Somehow solving YA is considered as a subproblem of solving ZA.  Even though ZA succeeds, the failure of YA somehow makes Agda think the original problem (solving ZA) fails.

**Comment by UlfNorell** [comment]
This is a little subtle. Suppose you have an unsolved constraint `C` (in the example `Y unit`) and are looking at a candidate instance solution `t : D`. Now suppose `C` fails after solving the instance problem with `t : D`. There are two scenarios:
- `C` fails for unrelated reasons (like in the example above), in which case we want to report a failure to solve `C`
- `C` fails because we solved the instance problem with `t`, in which case we'd like to discard `t` and pick a different solution (if any).

Currently we have no good way to distinguish between the two cases. Previous to the fix of #1915 we treated every constraint as being unrelated, whereas now we treat them as related. This is a deliberate choice and not a regression.
